thorlcr/master/thgraphmanager.cpp

```diff
@@ -1109,31 +1111,89 @@ void thorMain(ILogMsgHandler *logHandler, const char *wuid, const char *graphNam
-                IException *e = jobManager->queryExitException();
-                Owned<IWorkUnit> w = &workunit->lock();
-                if (e)
+                PROGLOG("Executing: wuid=%s, graph=%s", wuid, currentGraphName.str());
+
                 {
-                    Owned<IWUException> we = w->createException();
-                    we->setSeverity(SeverityInformation);
-                    StringBuffer errStr;
-                    e->errorMessage(errStr);
-                    we->setExceptionMessage(errStr);
-                    we->setExceptionSource("thormasterexception");
-                    we->setExceptionCode(e->errorCode());
+                    Owned<IWorkUnitFactory> factory;
+                    Owned<IConstWorkUnit> workunit;
+                    factory.setown(getWorkUnitFactory());
+                    workunit.setown(factory->openWorkUnit(wuid));
+                    SocketEndpoint dummyAgentEp;
+                    jobManager->execute(workunit, wuid, currentGraphName, dummyAgentEp);
+                    IException *e = jobManager->queryExitException();
+                    Owned<IWorkUnit> w = &workunit->lock();
+                    if (e)
+                    {
+                        Owned<IWUException> we = w->createException();
+                        we->setSeverity(SeverityInformation);
+                        StringBuffer errStr;
+                        e->errorMessage(errStr);
+                        we->setExceptionMessage(errStr);
+                        we->setExceptionSource("thormasterexception");
+                        we->setExceptionCode(e->errorCode());
+
+                        w->setState(WUStateWait);
+                        break;
+                    }
+
+                    if (lingerPeriod)
+                        w->setDebugValue(instance, "1", true);
+
+                    w->setState(WUStateWait);
                 }
-                w->setState(WUStateWait);
-            }  
-            else
-                jobManager->run();
+                currentGraphName.clear();
+
+                if (lingerPeriod)
+                {
+                    CMessageBuffer msg;
+                    StringBuffer next;
+                    CTimeMon timer(lingerPeriod);
+                    unsigned remaining;
+                    while (!timer.timedout(&remaining))
+                    {
+                        PROGLOG("Lingering time left: %.2f", ((float)remaining)/1000);
+                        if (!queryWorldCommunicator().recv(msg, NULL, MPTAG_THOR, nullptr, remaining))
+                            break;
+                        msg.read(next);
+
+                        // validate
+                        StringArray sArray;
+                        sArray.appendList(next, "/");
+                        if (2 == sArray.ordinality() && streq(sArray.item(0), wuid))
+                        {
+                            currentGraphName.set(sArray.item(1));
```

Comment:
Do we break out here if currentGraphName is empty ("") ?

Reply:
Yes see line 1193 below.

Reply:
Yes, sorry I read the break at line 1179 as part of the if on line 1177